beetsplug/absubmit.py

```diff
@@ -0,0 +1,159 @@
+            if not os.path.isfile(self.extractor):
+                raise FatalABSubmitError(
+                    u'extractor command does not exist: {0}'.
+                    format(self.extractor)
+                )
+        else:
+            # Implicit path to extractor, search for it in path
+            # TODO how to check for on Windows?
+            self.extractor = 'streaming_extractor_music'
+            try:
+                call([self.extractor])
+            except OSError:
+                raise FatalABSubmitError(
+                    u'no extractor command found: install "{0}"'.
+                    format(self.extractor)
+                )
+            except AnalysisABSubmitError:
+                # Extractor found, will exit with an error if not called with
+                # the correct amount of arguments.
+                pass
+            # Get the executable needed to calculate the sha1 hash.
+            self.extractor = distutils.spawn.find_executable(self.extractor)
+
+    supported_formats = {'mp3', 'ogg', 'oga', 'flac', 'mp4', 'm4a', 'm4r',
+                         'm4b', 'm4p', 'aac', 'wma', 'asf', 'mpc', 'wv',
+                         'spx', 'tta', '3g2', 'aif', 'aiff', 'ape'}
+
+    base_url = 'https://acousticbrainz.org/api/v1/{mbid}/low-level'
+
+    def commands(self):
+        cmd = ui.Subcommand(
+            'absubmit',
+            help=u'calculate and submit AcousticBrainz analysis'
+        )
+        cmd.func = self.command
+        return [cmd]
+
+    def command(self, lib, opts, args):
+        # Get items from arguments
+        items = lib.items(ui.decargs(args))
+        # Get no_submit option.
+        # TODO get a should submit option from the command line.
+        for item in items:
+            analysis = self._get_analysis(item)
+            if analysis:
+                self._submit_data(item, analysis)
+
+    def _get_analysis(self, item):
+        mbid = item['mb_trackid']
+        # If file has no mbid skip it.
+        if not mbid:
+            self._log.info('Not analysing {}, missing '
+                           'musicbrainz track id.', item)
+            return None
+        # If file format is not supported skip it.
+        if item['format'].lower() not in self.supported_formats:
+            self._log.info('Not analysing {}, file not in '
+                           'supported format.', item)
+            return None
+
+        # Temporary file to save extractor output to.
+        tmp_file, filename = tempfile.mkstemp(suffix='.json')
+        try:
+            # Close the file, so the extractor can overwrite it.
+            call([self.extractor, util.syspath(item.path), filename])
+            with open(filename) as tmp_file:
+                analysis = json.loads(tmp_file.read())
+            # Calculate extractor hash.
+            m = hashlib.sha1()
+            with open(self.extractor, 'rb') as extractor:
+                m.update(extractor.read())
```

Comment:
Yes, moving that shouldn't be a problem. Would you want it at init, or when the submit function is called?